snuba/request/request_settings.py

```diff
@@ -1,10 +1,69 @@
+from __future__ import annotations
+
 from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
 
-from typing import Sequence
+from typing import Any, Mapping, Optional, Sequence
 
 from snuba.state.rate_limit import get_global_rate_limit_params, RateLimitParameters
 
 
+class SamplingMode(Enum):
+    """
+    The value of this ennum are part of the public schema. Do not change them
+    without ensuring backward compatibility.
+    """
+
+    # No sampling applied.
+    NONE = "none"
+    # A fixed 0 - 1.0 (both exclusive) sampling rate is provided.
+    FIXED = "fixed"
+    # Adaptive sampling rate with aggregation adjustment. In this mode Snuba
+    # tries to guess an appropriate sampling rate depending on several heuristics
+    # including stats on the expected amount of data to query.
+    # Snuba will try its best to adjust the result of aggregations according to
+    # the sampling rate where it makes sense. If an aggregation function is passed
+    # that Snuba does not know how to adjust, Snuba will bail and apply no sampling
+    # so the client will still receive data that does not need adjustment, thus
+    # copmliant with what was requested.
+    ADAPTIVE_ADJUST = "adaptive_adjust"
+    # Adaptive sampling rate without aggregation adjustment. It works like for
+    # adaptive_adjust but Snuba will not apply any adjustment. It is responsibility
+    # of the client to apply the appropriate adjustment.
+    # This gives higher guarantees that sampling will actually be applied since
+    # Snuba does not need to know how to adjust the aggregations provided in the
+    # query.
+    ADAPTIVE_NO_ADJUST = "adaptive_no_adjust"
+    # The old automatic approach that applies a fixed config sampling if in TURBO
+    # mode.
+    AUTO_DEPRECATED = "auto"
+
+
+@dataclass(frozen=True)
+class SamplingConfig:
+    mode: SamplingMode
+    rate: Optional[float]
+
+    def __post_init__(self) -> None:
+        if self.mode != SamplingMode.FIXED:
+            assert (
+                self.rate is None
+            ), f"Sampling mode {self.mode} does not support a fixed rate"
+        else:
+            assert (
+                self.rate is not None
+            ), f"Sampling mode {self.mode} requires a fixed rate"
```

Comment:
This behavior seems to imply that the sampling modes should possibly be implemented as objects (some of which take a rate parameter and some of which don't) instead of trying to wrap them all in the same type.